openedx_learning/apps/authoring/publishing/models/draft_log.py

```diff
@@ -0,0 +1,239 @@
+
+
+class Draft(models.Model):
+    """
+    Find the active draft version of an entity (usually most recently created).
+
+    This model mostly only exists to allow us to join against a bunch of
+    PublishableEntity objects at once and get all their latest drafts. You might
+    use this together with Published in order to see which Drafts haven't been
+    published yet.
+
+    A Draft entry should be created whenever a new PublishableEntityVersion is
+    created. This means there are three possible states:
+
+    1. No Draft entry for a PublishableEntity: This means a PublishableEntity
+       was created, but no PublishableEntityVersion was ever made for it, so
+       there was never a Draft version.
+    2. A Draft entry exists and points to a PublishableEntityVersion: This is
+       the most common state.
+    3. A Draft entry exists and points to a null version: This means a version
+       used to be the draft, but it's been functionally "deleted". The versions
+       still exist in our history, but we're done using it.
+
+    It would have saved a little space to add this data to the Published model
+    (and possibly call the combined model something else). Split Modulestore did
+    this with its active_versions table. I keep it separate here to get a better
+    separation of lifecycle events: i.e. this table *only* changes when drafts
+    are updated, not when publishing happens. The Published model only changes
+    when something is published.
+    """
+    # If we're removing a PublishableEntity entirely, also remove the Draft
+    # entry for it. This isn't a normal operation, but can happen if you're
+    # deleting an entire LearningPackage.
+    entity = models.OneToOneField(
+        PublishableEntity,
+        on_delete=models.CASCADE,
+        primary_key=True,
+    )
+    version = models.OneToOneField(
+        PublishableEntityVersion,
+        on_delete=models.RESTRICT,
+        null=True,
+        blank=True,
+    )
+
+
+class DraftChangeLog(models.Model):
+    """
+    There is one row in this table for every time Drafts are created/modified.
+
+    Most of the time we'll only be changing one Draft at a time, and this will
+    be 1:1 with DraftChange. But there are some operations that affect many
+    Drafts at once, such as discarding changes (i.e. reset to the published
+    version) or doing an import.
+    """
+    uuid = immutable_uuid_field()
+    learning_package = models.ForeignKey(LearningPackage, on_delete=models.CASCADE)
+    changed_at = manual_date_time_field()
+    changed_by = models.ForeignKey(
+        settings.AUTH_USER_MODEL,
+        on_delete=models.SET_NULL,
+        null=True,
+        blank=True,
+    )
+
+    @property
+    def changes(self):
+        """
+        Alias for "records".
+
+        We use "records" for consistency with PublishLogRecord, but somteimes
+        it's more natural to call these "changes".
+        """
+        return self.records
```

Comment:
I'm not big on this. In trying to have both consistency (records) and clarity (changes), we'll end up with neither.

Reply:
Fair. I'll get rid of this alias and use `records` everywhere.